File: src/tools/onboarding.ts
import { type ToolDefinition, type ToolResult } from "../core/types.js";

export function getTools(): ToolDefinition[] {
  const make = (n: string, d: string, s: Record<string, unknown>, fn: (p: any, ctx: any) => Promise<ToolResult>): ToolDefinition => ({
    name: n, description: d, inputSchema: s as any, execute: fn,
  });
  return [
    make("onboarding_state", "Get first-boot state. Read-only.", { type:"object", properties:{} },
      async (_p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("query { onboarding { state } }") })),
    make("onboarding_complete", "Complete onboarding.", { type:"object", properties:{} },
      async (_p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("mutation { completeOnboarding { status } }") })),
    make("onboarding_open", "Open onboarding.", { type:"object", properties:{} },
      async (_p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("mutation { openOnboarding { status } }") })),
    make("onboarding_close", "Close onboarding.", { type:"object", properties:{} },
      async (_p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("mutation { closeOnboarding { status } }") })),
    make("onboarding_resume", "Resume onboarding.", { type:"object", properties:{} },
      async (_p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("mutation { resumeOnboarding { status } }") })),
    make("onboarding_bypass", "Bypass onboarding.", { type:"object", properties:{} },
      async (_p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("mutation { bypassOnboarding { status } }") })),
    make("onboarding_reset", "Reset onboarding.", { type:"object", properties:{ confirm:{type:"boolean"} }, required:["confirm"] },
      async (p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("mutation { resetOnboarding { status } }", p) })),
    make("onboarding_create_internal_boot_pool", "Create boot pool.", { type:"object", properties:{} },
      async (_p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("mutation { createInternalBootPool { status } }") })),
  ];
}

File: go/internal/tools/onboarding.go
package onboarding

import (
	"context"
	"fmt"

	"github.com/hektyc/unraid-mcp-server/go/internal/config"
	"github.com/hektyc/unraid-mcp-server/go/internal/guards"
	"github.com/hektyc/unraid-mcp-server/go/internal/mcp"
)

func RegisterTools(s *mcp.Server, cfg *config.Config) {
	s.RegisterTool(&mcp.ToolDef{
		Name:        "onboarding_state",
		Description: "Get first-boot state. Read-only.",
		Query:       `query { onboarding { state } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "onboarding_complete",
		Description: "Complete onboarding.",
		Query:       `mutation { completeOnboarding { status } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "onboarding_reset",
		Description: "Reset onboarding.",
		Query:       `mutation { resetOnboarding { status } }`,
		Params: map[string]string{
			"confirm": "boolean", // required=true
		},
	})
}